File: ballast-core/npm/src/domain/models/game.ts
import { Board } from "./board";
import { Player } from "./player";
import { Vessel } from "./vessel";
import { VesselRole } from "./vessel-role";
import { CubicCoordinates } from "./cubic-coordinates";

export class Game {

    public readonly id: string;
    public readonly board: Board;
    public readonly vessels: Vessel[];
    public readonly players: Player[];
    public createdOnDate: Date;
    public startedOnDate: Date | null;
    public endedOnDate: Date | null;

    public constructor(
        id: string, board: Board, vessels: Vessel[], players: Player[], 
        createdOnDate: Date, startedOnDate: Date | null = null, endedOnDate: Date | null = null
    ) {
        this.id = id;
        this.board = board;
        this.vessels = vessels.slice(0);
        this.players = players.slice(0);
        this.createdOnDate = createdOnDate
        this.startedOnDate = startedOnDate || null;
        this.endedOnDate = endedOnDate || null;
    }

    public updateVesselCoordinates(vesselId: string, cubicCoordinates: CubicCoordinates)
    {
        var foundVessel = this.vessels.find(x => x.id == vesselId);
        if (!foundVessel)
            throw new Error(`Could not locate vessel with id '${vesselId}'`);
        var foundTile = this.board.getTileFromCoordinates(cubicCoordinates);
        if (!foundTile)
            throw new Error(`Could not locate tile with coordinates ${cubicCoordinates.toOrderedTriple()}`);
        if (!foundTile.terrain.passable)
            throw new Error(`The requested tile terrain is not passable`);
        return foundVessel.updateCoordinates(cubicCoordinates);
    }

    public start(): Date {
        if (!!this.startedOnDate)
            throw new Error("Can't re-start game after it has already begun!");
        var startedOnDate = new Date(Date.now());
        this.startedOnDate = startedOnDate;
        return startedOnDate;
    }

    public end(): Date {
        if (!!this.endedOnDate)
            throw new Error("Can't end game after it has already finished!");
        var endedOnDate = new Date(Date.now());
        this.endedOnDate = endedOnDate;
        return endedOnDate;
    }

    public addPlayer(player: Player): Player {
        if (!player || !player.id)
            throw new Error(`Could not locate player matching id ${player && player.id}`);
        if (this.players.find(x => x.id == player.id))
            throw new Error(`Player with id ${player.id} already exists`);
        this.players.push(player);
        return player;
    }

    public setVesselRole(vesselId: string, vesselRole : VesselRole, player: Player) {
        var vessel = this.vessels.find(x => x.id == vesselId);
        if (!vessel)
            throw new Error(`Could not find vessel for id ${vesselId}`);
        vessel.setVesselRole(vesselRole, player);
    }

    public removePlayerById(playerId: string) {
        if (!playerId)
            throw new Error("Player id is required");
        var player = this.players.find(x => x.id == playerId);
        if (!!player) {
            this.removePlayer(player);
        }
    }

    public removePlayer(player: Player) {
        if (!player || !player.id)
            throw new Error("Player.id is required");
        for(let vessel of this.vessels)
        {
           vessel.removePlayer(player); 
        }
        var playerIndex = this.players.findIndex(x => x.id == player.id);
        if (playerIndex > 0)
            this.players.splice(playerIndex, 1);
    }

}

File: ballast-core/dotnet/src/domain/models/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Ballast.Core.Domain.Models
{
    public class Game
    {

        public Guid Id { get; private set; }
        public Board Board { get; private set; }
        public DateTime CreatedOnDate { get; private set; }
        public DateTime? StartedOnDate { get; private set; }
        public DateTime? EndedOnDate { get; private set; }

        public IEnumerable<Vessel> Vessels => _vessels;
        private IList<Vessel> _vessels;

        public IEnumerable<Player> Players => _players;
        private IList<Player> _players;

        public Game(
            Guid id, 
            Board board, 
            IEnumerable<Vessel> vessels, 
            IEnumerable<Player> players,
            DateTime? createdOnDate = null,
            DateTime? startedOnDate = null,
            DateTime? endedOnDate = null
        )
        {
            Id = id;
            Board = board;
            _vessels = vessels.ToList();
            _players = players.ToList();
            CreatedOnDate = createdOnDate ?? DateTime.UtcNow;
            StartedOnDate = startedOnDate;
            EndedOnDate = endedOnDate;
        }
        
        public CubicCoordinates UpdateVesselCoordinates(Guid vesselId, CubicCoordinates cubicCoordinates)
        {
            var foundVessel = this._vessels.SingleOrDefault(x => x.Id == vesselId);
            if (foundVessel == null)
                throw new KeyNotFoundException($"Could not locate vessel with id '{vesselId}'");
            var foundTile = this.Board.GetTileFromCoordinates(cubicCoordinates);
            if (foundTile == null) 
                throw new KeyNotFoundException($"Could not locate tile with coordinates '{cubicCoordinates.ToOrderedTriple()}'");
            if (!foundTile.Terrain.Passable)
                throw new InvalidOperationException($"The requested tile terrain is not passable");
            return foundVessel.UpdateCoordinates(cubicCoordinates);
        }

        public DateTime Start()
        {
            if (StartedOnDate != null)
                throw new InvalidOperationException("Can't re-start game after it has already begun!");
            var startedOnDate = DateTime.UtcNow;
            StartedOnDate = startedOnDate;
            return startedOnDate;
        }

        public DateTime End()
        {
            if (EndedOnDate != null)
                throw new InvalidOperationException("Can't end game after it has already finished!");
            var endedOnDate = DateTime.UtcNow;
            EndedOnDate = endedOnDate;
            return endedOnDate;
        }

        public Player AddPlayer(Player player) 
        {
            if (player == null || player.Id == default(Guid))
                throw new ArgumentNullException(nameof(player.Id));
            if (_players.Any(x => x.Id == player.Id))
                throw new ArgumentException($"Player with id {player?.Id} already exists");
            _players.Add(player);
            return player;
        }

        public void SetVesselRole(Guid vesselId, VesselRole vesselRole, Player player)
        {
            var vessel = _vessels.FirstOrDefault(x => x.Id == vesselId);
            if (vessel == null)
                throw new KeyNotFoundException($"Could not find vessel for id {vesselId}");
            vessel.SetVesselRole(vesselRole, player);
        }

        public void RemovePlayerById(Guid playerId) 
        {
            if (playerId == default(Guid))
                throw new ArgumentNullException(nameof(playerId));
            var player = _players.FirstOrDefault(x => x.Id == playerId);
            RemovePlayer(player);
        }

        public void RemovePlayer(Player player) 
        {
            if (player == null || player.Id == default(Guid))
                throw new ArgumentNullException(nameof(player.Id));
            foreach(var vessel in _vessels)
            {
               vessel.RemovePlayer(player); 
            }
            if (_players.Any(x => x.Id == player.Id))
                _players.Remove(player);
        }
        
    }
}